geonode/api/tests.py

```diff
@@ -226,3 +236,222 @@ def test_date_filter(self):
+        resp = self.api_client.get(filter_url)
+        self.assertValidJSONResponse(resp)
+        # This is a list url
+        objects = self.deserialize(resp)['objects']
+
+        self.assertEqual(len(objects), 1)
+
+        # Check necessary list fields
+        obj = objects[0]
+        field_list = [
+            'layer',
+            'name',
+            'title',
+            'style_url',
+            'type',
+            'resource_uri'
+        ]
+
+        # Additional field based on OGC Backend
+        if check_ogc_backend(geoserver.BACKEND_PACKAGE):
+            field_list += [
+                'version',
+                'workspace'
+            ]
+        elif check_ogc_backend(qgis_server.BACKEND_PACKAGE):
+            field_list += [
+                'style_legend_url'
+            ]
+        for f in field_list:
+            self.assertTrue(f in obj)
+
+        if check_ogc_backend(geoserver.BACKEND_PACKAGE):
+            self.assertEqual(obj['type'], 'sld')
+        elif check_ogc_backend(qgis_server.BACKEND_PACKAGE):
+            self.assertEqual(obj['type'], 'qml')
+
+        # Check style detail
+        detail_url = obj['resource_uri']
+        resp = self.api_client.get(detail_url)
+        self.assertValidJSONResponse(resp)
+        obj = self.deserialize(resp)
+
+        # should include body field
+        self.assertTrue('body' in obj and obj['body'])
+
+    def test_add_delete_styles(self):
+        """Style API Add/Delete interaction."""
+        # Check styles count
+        style_list_url = reverse(
+            'api_dispatch_list',
+            kwargs={
+                'api_name': 'api',
+                'resource_name': 'styles'
+            }
+        )
+        filter_url = style_list_url + '?layer__name=' + self.layer.name
+        resp = self.api_client.get(filter_url)
+        self.assertValidJSONResponse(resp)
+        objects = self.deserialize(resp)['objects']
+
+        self.assertEqual(len(objects), 1)
+
+        # Fetch default style
+        layer_detail_url = reverse(
+            'api_dispatch_detail',
+            kwargs={
+                'api_name': 'api',
+                'resource_name': 'layers',
+                'id': self.layer.id
+            }
+        )
+        resp = self.api_client.get(layer_detail_url)
+        self.assertValidJSONResponse(resp)
+        obj = self.deserialize(resp)
+        # Take default style url from Layer detail info
+        default_style_url = obj['default_style']
+        resp = self.api_client.get(default_style_url)
+        self.assertValidJSONResponse(resp)
+        obj = self.deserialize(resp)
+        style_body = obj['body']
+
+        style_stream = StringIO(style_body)
+        # Add virtual filename
+        style_stream.name = 'style.qml'
+        data = {
+            'layer__id': self.layer.id,
+            'name': 'new_style',
+            'title': 'New Style',
+            'style': style_stream
+        }
```

Comment:
Above is a necessary data to post a new QML style.
The format is ```multipart/form-data```
`style` field is a FILE data.
Other are POST data.